File: src/structure/namespace/createNamespaceInfoItem.ts
import ts, {ModuleDeclaration} from "typescript";
import {Configuration, NamespaceStrategy} from "../../configuration/configuration.js";
import {StructureItem} from "../structure.js";
import {moduleNameToPackage} from "../module/moduleNameToPackage.js";
import {extractModuleName} from "../module/extractModuleName.js";

export interface NamespaceInfoItem extends StructureItem {
    name: string
    strategy: NamespaceStrategy
}

interface NamespaceNameChunk {
    detailedName: string
    simpleName: string
    package: string[]
    isAmbient: boolean
}

export const defaultNamespaceStrategy: NamespaceStrategy = "object"

export function extractNamespaceName(
    namespace: ModuleDeclaration,
    suffix: NamespaceNameChunk[] = []
): NamespaceNameChunk[] {
    const simpleName = namespace.name.text
    const detailedName = ts.isIdentifier(namespace.name)
        ? simpleName
        : `"${simpleName}"`
    const packageChunks = moduleNameToPackage(simpleName).map(it => it.toLowerCase())
    const isAmbient = ts.isStringLiteral(namespace.name)

    const nameChunk = {
        detailedName,
        simpleName,
        package: packageChunks,
        isAmbient,
    }

    const name = [nameChunk, ...suffix]

    if (ts.isModuleDeclaration(namespace.parent)) {
        return extractNamespaceName(namespace.parent, name)
    } else if (ts.isModuleBlock(namespace.parent)) {
        return extractNamespaceName(namespace.parent.parent, name)
    } else {
        return name
    }
}

export function createNamespaceInfoItem(
    namespace: ModuleDeclaration,
    sourceFileName: string,
    imports: string[],
    configuration: Configuration,
): NamespaceInfoItem {
    const {namespaceStrategy} = configuration

    const defaultModuleName = extractModuleName(sourceFileName, configuration)
    const name = extractNamespaceName(namespace)

    const detailedName = name.map(it => it.detailedName).join(".")
    const simpleName = name.map(it => it.simpleName).join(".")

    const fileName = "namespace.kt"
    const packageChunks = name.flatMap(it => it.package)
    const hasRuntime = true

    let moduleName = defaultModuleName
    let qualifier: string | undefined = simpleName

    const [firstChunk, ...restChunks] = name

    if (firstChunk.isAmbient) {
        moduleName = firstChunk.simpleName

        if (restChunks.length > 0) {
            qualifier = restChunks.map(it => it.simpleName).join(".")
        } else {
            qualifier = undefined
        }
    }

    for (const [pattern, strategy] of Object.entries(namespaceStrategy)) {
        const regexp = new RegExp(pattern)

        if (regexp.test(detailedName) || regexp.test(simpleName)) {
            return {
                fileName,
                package: packageChunks,
                moduleName,
                qualifier,
                hasRuntime,
                imports,
                name: detailedName,
                strategy
            }
        }
    }

    return {
        fileName,
        package: packageChunks,
        moduleName,
        qualifier,
        hasRuntime,
        imports,
        name: detailedName,
        strategy: defaultNamespaceStrategy
    }
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/structure/namespace/createNamespaceInfoItem.kt
package io.github.sgrishchenko.karakum.structure.namespace

import io.github.sgrishchenko.karakum.configuration.Configuration
import io.github.sgrishchenko.karakum.configuration.NamespaceStrategy
import io.github.sgrishchenko.karakum.structure.StructureItem
import io.github.sgrishchenko.karakum.structure.module.extractModuleName
import io.github.sgrishchenko.karakum.structure.module.moduleNameToPackage
import js.array.ReadonlyArray
import js.objects.JsPlainObject
import js.objects.Object
import typescript.*

@JsPlainObject
external interface NamespaceInfoItem : StructureItem {
    val name: String
    val strategy: NamespaceStrategy
}

@JsPlainObject
external interface NamespaceNameChunk {
    val detailedName: String
    val simpleName: String
    val `package`: ReadonlyArray<String>
    val isAmbient: Boolean
}

val defaultNamespaceStrategy = NamespaceStrategy.`object`

fun extractNamespaceName(
    namespace: ModuleDeclaration,
    suffix: ReadonlyArray<NamespaceNameChunk> = emptyArray()
): ReadonlyArray<NamespaceNameChunk> {
    val namespaceName = namespace.name
    var simpleName = ""

    if (isStringLiteral(namespaceName)) simpleName = namespaceName.text
    if (isIdentifier(namespaceName)) simpleName = namespaceName.text
    if (simpleName.isEmpty()) error("Unsupported namespace name")

    val detailedName =
        if (isIdentifier(namespaceName)) simpleName
        else "\"${simpleName}\""

    val packageChunks = moduleNameToPackage(simpleName).map { it.lowercase() }
    val isAmbient = isStringLiteral(namespace.name)

    val nameChunk = NamespaceNameChunk(
        detailedName = detailedName,
        simpleName = simpleName,
        `package` = packageChunks.toTypedArray(),
        isAmbient = isAmbient,
    )

    val name = arrayOf(nameChunk) + suffix
    val parent = namespace.parent

    return if (isModuleDeclaration(parent)) {
        extractNamespaceName(parent, name)
    } else if (isModuleBlock(parent)) {
        extractNamespaceName(parent.parent, name)
    } else {
        name
    }
}

fun createNamespaceInfoItem(
    namespace: ModuleDeclaration,
    sourceFileName: String,
    imports: ReadonlyArray<String>,
    configuration: Configuration,
): NamespaceInfoItem {
    val namespaceStrategy = configuration.namespaceStrategy

    val defaultModuleName = extractModuleName(sourceFileName, configuration)
    val name = extractNamespaceName(namespace)

    val detailedName = name.joinToString(separator = ".") { it.detailedName }
    val simpleName = name.joinToString(separator = ".") { it.simpleName }

    val fileName = "namespace.kt"
    val packageChunks = name.flatMap { it.`package`.toList() }
    val hasRuntime = true

    var moduleName = defaultModuleName
    var qualifier: String? = simpleName

    val (firstChunk, restChunks) = with(name) { first() to drop(1) }

    if (firstChunk.isAmbient) {
        moduleName = firstChunk.simpleName

        qualifier = if (restChunks.isNotEmpty()) {
            restChunks.joinToString(separator = ".") { it.simpleName }
        } else {
            null
        }
    }

    for ((pattern, strategy) in Object.entries(namespaceStrategy)) {
        val regexp = pattern.toRegex()

        if (regexp.containsMatchIn(detailedName) || regexp.containsMatchIn(simpleName)) {
            return NamespaceInfoItem(
                fileName = fileName,
                `package` = packageChunks.toTypedArray(),
                moduleName = moduleName,
                qualifier = qualifier,
                hasRuntime = hasRuntime,
                imports = imports,
                name = detailedName,
                strategy = strategy
            )
        }
    }

    return NamespaceInfoItem(
        fileName = fileName,
        `package` = packageChunks.toTypedArray(),
        moduleName = moduleName,
        qualifier = qualifier,
        hasRuntime = hasRuntime,
        imports = imports,
        name = detailedName,
        strategy = defaultNamespaceStrategy
    )
}
